### PR TITLE
Added `allow_merges` option for volumes in multi-disk configuration

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog

Reverted in https://github.com/ClickHouse/ClickHouse/pull/13954

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Added `allow_merges` option for volumes in multi-disk configuration.

### CODE REVIEW

**Comment by Akazz** [comment]
@excitoon This PR lacks tests to say the least. You should definitely add some stateless tests

File: src/Storages/MergeTree/TTLMergeSelector.cpp [hunk 1]
```diff
@@ -73,12 +80,14 @@ IMergeSelector::PartsInPartition TTLMergeSelector::select(
         --best_begin;
     }
 
+    /// Find end of range with most old TTL.
     while (best_end != best_partition.end())
```
Comment (Akazz) [inline src/Storages/MergeTree/TTLMergeSelector.cpp:84]: Hmm, what if we are at the end of parts already? (meaning we only have a single part selected by TTLMergeSelector) Is it okay to return 1 part?

File: src/Disks/VolumeJBOD.h [hunk 2]
```diff
@@ -38,11 +55,18 @@ class VolumeJBOD : public IVolume
...
     ReservationPtr reserve(UInt64 bytes) override;
 
+    bool areMergesAllowed() const override;
+
+    void setAllowMergesFromQuery(bool allow) override;
+
+    /// True if parts on this volume participate in merges according to configuration.
+    bool are_merges_allowed_in_config = true;
+
+    /// True if parts on this volume participate in merges according to START/STOP MERGES ON VOLUME.
```
Comment (Akazz) [inline src/Disks/VolumeJBOD.h:65]: I do not quite understand why we need to manage this flag separately from `are_merges_allowed_in_config`. Could you please comment on this?

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 3]
```diff
@@ -676,6 +676,16 @@ void IMergeTreeDataPart::loadColumns(bool require)
...
 }
 
+bool IMergeTreeDataPart::areMergesAllowed() const
+{
+    auto storage_policy = storage.getStoragePolicy();
+
+    /// One need to get volume description for given volume.
+    auto volume_ptr = storage_policy->getVolume(storage_policy->getVolumeIndexByDisk(volume->getDisk()));
+
+    return volume_ptr->areMergesAllowed();
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/Storages/MergeTree/IMergeTreeDataPart.cpp:686]: Sure!

[the remark this answers is not in the document]
Comment (excitoon) [reply src/Storages/MergeTree/IMergeTreeDataPart.cpp:686]: Actually I had the comment but not much clear one.

[the remark this answers is not in the document]
Comment (excitoon) [reply src/Storages/MergeTree/IMergeTreeDataPart.cpp:686]: +

File: src/Storages/MergeTree/IMergeTreeDataPart.h [hunk 4]
```diff
@@ -324,6 +324,9 @@ class IMergeTreeDataPart : public std::enable_shared_from_this<IMergeTreeDataPar
     /// storage and pass it to this method.
     virtual bool hasColumnFiles(const String & /* column */, const IDataType & /* type */) const{ return false; }
 
+    /// Checks if there any objections to merge this part on per-part basis.
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/Storages/MergeTree/IMergeTreeDataPart.h:327]: +

File: src/Storages/MergeTree/TTLMergeSelector.cpp (hunk 1 above)
Comment (excitoon) [reply src/Storages/MergeTree/TTLMergeSelector.cpp:84]: Yes, it is okay, TTL merges can occur with 1 part (for 'thinning', if I used correct word for than).

File: src/Disks/VolumeJBOD.h (hunk 2 above)
Comment (excitoon) [reply src/Disks/VolumeJBOD.h:65]: This setting comes from two different sources, configuration (which can be reloaded) and user will (START / STOP). These two flags simplify conflict management of these two sources.

**Comment by Akazz** [comment]
The failing test `1193_metadata_loading` in "Functional stateless tests (release, DatabaseAtomic)" looks irrelevant to the changes in this PR

**Comment by excitoon** [comment]
Progress on #10650

**Review by Akazz — APPROVED** [review]
LGTM
Yet, I suggest that you rename `allow_merges` parameter to what has been proposed in the original issue - `disable_merges`, as it better reflects what the setting does.
That can be done in a separate PR, of course.

**Pull request merged**